libraries/chain/webassembly/privileged.cpp

```diff
@@ -152,28 +153,30 @@ namespace eosio { namespace chain { namespace webassembly {
    }
 
    void interface::set_finalizers(span<const char> packed_finalizer_set) {
-      EOS_ASSERT(!context.trx_context.is_read_only(), wasm_execution_error, "set_proposed_finalizers not allowed in a readonly transaction");
+      EOS_ASSERT(!context.trx_context.is_read_only(), wasm_execution_error, "set_finalizers not allowed in a readonly transaction");
       fc::datastream<const char*> ds( packed_finalizer_set.data(), packed_finalizer_set.size() );
       finalizer_set finset;
-      fc::raw::unpack(ds, finset);
-      vector<finalizer_authority> & finalizers = finset.finalizers;
+      // contract finalizer_set does not include uint32_t generation
+      // struct abi_finalizer_set {
+      //   uint64_t fthreshold
+      //   vector<finalizer_authority> finalizers; }
+      fc::raw::unpack(ds, finset.fthreshold);
+      fc::raw::unpack(ds, finset.finalizers);
+
+      vector<finalizer_authority>& finalizers = finset.finalizers;
 
-      // TODO: check version and increment it or verify correct
       EOS_ASSERT( finalizers.size() <= config::max_finalizers, wasm_execution_error, "Finalizer set exceeds the maximum finalizer count for this chain" );
       EOS_ASSERT( finalizers.size() > 0, wasm_execution_error, "Finalizer set cannot be empty" );
 
       std::set<fc::crypto::blslib::bls_public_key> unique_finalizer_keys;
-#warning REVIEW: Is checking for unique finalizer descriptions at all relevant?
-      std::set<std::string> unique_finalizers;
       uint64_t f_weight_sum = 0;
 
       for (const auto& f: finalizers) {
+         EOS_ASSERT( f.description.size() <= config::max_finalizer_description, wasm_execution_error, "Finalizer description greater than 256" );
          f_weight_sum += f.fweight;
          unique_finalizer_keys.insert(f.public_key);
-         unique_finalizers.insert(f.description);
       }
 
-      EOS_ASSERT( finalizers.size() == unique_finalizers.size(), wasm_execution_error, "Duplicate finalizer description in finalizer set" );
       EOS_ASSERT( finalizers.size() == unique_finalizer_keys.size(), wasm_execution_error, "Duplicate finalizer bls key in finalizer set" );
       EOS_ASSERT( finset.fthreshold > f_weight_sum / 2, wasm_execution_error, "Finalizer set threshold cannot be met by finalizer weights" );
```

Comment:
why `/ 2`?

Reply:
Well, the spec mentions 15, which is 2/3+1 of 21.  I wonder if we should not check that `threshold > (f_weight_sum / 3) * 2`?

Reply:
I agree, seems like it should be `threshold >= (f_weight_sum * 2 / 3) + 1`  or `threshold > (f_weight_sum * 2 / 3)` .

Reply:
Conversation summary: The host function will enforce `f_weight_sum / 2`. It will be up to system contract to enforce a higher threshold if desired. 

> So if the chain would prefer to be more secure against finality violations at the cost of increased risk of losing liveness, it can increase the threshold above 1/2. Perhaps even all the way to 100% if it wanted.